File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/play_queue/PlayQueueDaoWrapper.java
package com.github.anrimian.musicplayer.data.database.dao.play_queue;

import com.github.anrimian.musicplayer.data.database.AppDatabase;
import com.github.anrimian.musicplayer.data.database.entities.play_queue.PlayQueueEntity;
import com.github.anrimian.musicplayer.data.database.entities.play_queue.PlayQueueItemDto;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem;
import com.github.anrimian.musicplayer.domain.utils.functions.Optional;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Observable;

import static com.github.anrimian.musicplayer.data.repositories.state.UiStateRepositoryImpl.NO_ITEM;
import static com.github.anrimian.musicplayer.domain.Constants.NO_POSITION;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.mapList;

/**
 * Created on 02.07.2018.
 */
public class PlayQueueDaoWrapper {

    private final AppDatabase appDatabase;
    private final PlayQueueDao playQueueDao;

    @Nullable
    private PlayQueueEntity deletedItem;

    public PlayQueueDaoWrapper(AppDatabase appDatabase, PlayQueueDao playQueueDao) {
        this.appDatabase = appDatabase;
        this.playQueueDao = playQueueDao;
    }

    public Observable<List<PlayQueueItem>> getPlayQueueObservable(boolean isRandom) {
        return isRandom? getPlayQueueInShuffledOrderObservable(): getPlayQueueInNormalOrderObservable();
    }

    public void reshuffleQueue(long currentItemId) {
        appDatabase.runInTransaction(() -> {
            List<PlayQueueEntity> list = playQueueDao.getPlayQueue();
            if (list.isEmpty()) {
                return;
            }

            Collections.shuffle(list);

            long firstItemId = list.get(0).getId();
            int currentItemPosition = -1;
            for (int i = 0; i < list.size(); i++) {
                PlayQueueEntity entity = list.get(i);

                if (entity.getId() == currentItemId) {
                    currentItemPosition = i;
                }
                entity.setShuffledPosition(i);
            }
            if (currentItemPosition != -1 && firstItemId != currentItemId) {
                list.get(currentItemPosition).setShuffledPosition(0);
                list.get(0).setShuffledPosition(currentItemPosition);
            }

            playQueueDao.deletePlayQueue();
            playQueueDao.insertItems(list);
        });
    }

    public long insertNewPlayQueue(List<Composition> compositions,
                                   boolean randomPlayingEnabled,
                                   int startPosition) {
        return appDatabase.runInTransaction(() -> {
            List<Composition> shuffledList = new ArrayList<>(compositions);
            long randomSeed = System.nanoTime();
            Collections.shuffle(shuffledList, new Random(randomSeed));

            List<Integer> shuffledPositionList = new ArrayList<>(compositions.size());
            for (int i = 0; i < compositions.size(); i++) {
                shuffledPositionList.add(i);
            }
            Collections.shuffle(shuffledPositionList, new Random(randomSeed));

            List<PlayQueueEntity> entities = new ArrayList<>(compositions.size());
            int shuffledStartPosition = 0;
            for (int i = 0; i < compositions.size(); i++) {
                Composition composition = compositions.get(i);
                PlayQueueEntity playQueueEntity = new PlayQueueEntity();
                playQueueEntity.setAudioId(composition.getId());
                playQueueEntity.setPosition(i);
                int shuffledPosition =  shuffledPositionList.get(i);
                playQueueEntity.setShuffledPosition(shuffledPosition);

                if (startPosition != NO_POSITION && i == startPosition) {
                    shuffledStartPosition = shuffledPosition;
                }

                entities.add(playQueueEntity);
            }

            playQueueDao.deletePlayQueue();
            playQueueDao.insertItems(entities);

            if (randomPlayingEnabled) {
                return playQueueDao.getItemIdAtShuffledPosition(shuffledStartPosition);
            } else {
                return playQueueDao.getItemIdAtPosition(startPosition == NO_POSITION? 0: startPosition);
            }
        });
    }

    public Observable<Optional<PlayQueueItem>> getItemObservable(long id) {
        return playQueueDao.getItemObservable(id)
                .map(dto -> {
                    PlayQueueItem item = null;
                    if (dto.length > 0) {
                        item = toQueueItem(dto[0]);
                    }
                    return new Optional<>(item);
                });
    }

    public void deleteItem(long itemId) {
        deletedItem = playQueueDao.getItem(itemId);
        playQueueDao.deleteItem(itemId);
    }

    @Nullable
    public Long restoreDeletedItem() {
        if (deletedItem != null) {
           return playQueueDao.insertItem(deletedItem);
        }
        return null;
    }

    public void swapItems(PlayQueueItem firstItem, PlayQueueItem secondItem, boolean shuffleMode) {
        appDatabase.runInTransaction(() -> {
            long firstId = firstItem.getId();
            long secondId = secondItem.getId();
            if (shuffleMode) {
                int firstPosition = playQueueDao.getShuffledPosition(firstId);
                int secondPosition = playQueueDao.getShuffledPosition(secondId);

                playQueueDao.updateShuffledPosition(secondId, Integer.MIN_VALUE);
                playQueueDao.updateShuffledPosition(firstId, secondPosition);
                playQueueDao.updateShuffledPosition(secondId, firstPosition);
            } else {
                int firstPosition = playQueueDao.getPosition(firstId);
                int secondPosition = playQueueDao.getPosition(secondId);

                playQueueDao.updateItemPosition(secondId, Integer.MIN_VALUE);
                playQueueDao.updateItemPosition(firstId, secondPosition);
                playQueueDao.updateItemPosition(secondId, firstPosition);
            }
        });
    }

    public long addCompositionsToEndQueue(List<Composition> compositions) {
        return appDatabase.runInTransaction(() -> {
            int positionToInsert = playQueueDao.getLastPosition() + 1;
            int shuffledPositionToInsert = playQueueDao.getLastShuffledPosition() + 1;
            List<PlayQueueEntity> entities = toEntityList(compositions,
                    positionToInsert,
                    shuffledPositionToInsert);
            long[] ids = playQueueDao.insertItems(entities);
            return ids[0];
        });
    }

    public long addCompositionsToQueue(List<Composition> compositions, long currentItemId) {
        return appDatabase.runInTransaction(() -> {
            int positionToInsert = 0;
            int shuffledPositionToInsert = 0;
            if (currentItemId != NO_ITEM) {
                int currentPosition = playQueueDao.getPosition(currentItemId);
                int currentShuffledPosition = playQueueDao.getShuffledPosition(currentItemId);

                int increaseBy = compositions.size();
                int lastPosition = playQueueDao.getLastPosition();
                for (int pos = lastPosition; pos > currentPosition; pos--) {
                    playQueueDao.increasePosition(increaseBy, pos);
                }
                int lastShuffledPosition = playQueueDao.getLastShuffledPosition();
                for (int pos = lastShuffledPosition; pos > currentShuffledPosition; pos--) {
                    playQueueDao.increaseShuffledPosition(increaseBy, pos);
                }

                positionToInsert = currentPosition + 1;
                shuffledPositionToInsert = currentShuffledPosition + 1;
            }

            List<PlayQueueEntity> entities = toEntityList(compositions,
                    positionToInsert,
                    shuffledPositionToInsert);
            long[] ids = playQueueDao.insertItems(entities);
            return ids[0];
        });
    }

    public int getPosition(long id, boolean isShuffle) {
        if (isShuffle) {
            return playQueueDao.getShuffledPosition(id);
        } else {
            return playQueueDao.getPosition(id);
        }
    }

    public int getLastPosition(boolean isShuffled) {
        if (isShuffled) {
            return playQueueDao.getLastShuffledPosition();
        } else {
            return playQueueDao.getLastPosition();
        }
    }

    public Observable<Integer> getPositionObservable(long id, boolean isShuffle) {
        Observable<Integer> observable;
        if (isShuffle) {
            observable = playQueueDao.getShuffledPositionObservable(id);
        } else {
            observable = playQueueDao.getPositionObservable(id);
        }
        return observable.distinctUntilChanged();
    }

    public Observable<Integer> getIndexPositionObservable(long id, boolean isShuffle) {
        Observable<Integer> observable;
        if (isShuffle) {
            observable = playQueueDao.getShuffledIndexPositionObservable(id);
        } else {
            observable = playQueueDao.getIndexPositionObservable(id);
        }
        return observable.filter(pos -> pos >= 0)
                .distinctUntilChanged();
    }

    public long getNextQueueItemId(long currentItemId, boolean isShuffled) {
        if (isShuffled) {
            Long id = playQueueDao.getNextShuffledQueueItemId(currentItemId);
            if (id == null) {
                return playQueueDao.getFirstShuffledItem();
            }
            return id;
        } else {
            Long id = playQueueDao.getNextQueueItemId(currentItemId);
            if (id == null) {
                return playQueueDao.getFirstItem();
            }
            return id;
        }
    }

    public long getPreviousQueueItemId(long currentItemId, boolean isShuffled) {
        if (isShuffled) {
            Long id = playQueueDao.getPreviousShuffledQueueItemId(currentItemId);
            if (id == null) {
                return playQueueDao.getLastShuffledItem();
            }
            return id;
        } else {
            Long id = playQueueDao.getPreviousQueueItemId(currentItemId);
            if (id == null) {
                return playQueueDao.getLastItem();
            }
            return id;
        }
    }

    public Long getItemAtPosition(int position, boolean isShuffled) {
        if (isShuffled) {
            return playQueueDao.getItemIdAtShuffledPosition(position);
        } else {
            return playQueueDao.getItemIdAtPosition(position);
        }
    }

    public void deletePlayQueue() {
        playQueueDao.deletePlayQueue();
    }

    public Observable<Integer> getPlayQueueSizeObservable() {
        return playQueueDao.getPlayQueueSizeObservable();
    }

    private Observable<List<PlayQueueItem>> getPlayQueueInNormalOrderObservable() {
        return playQueueDao.getPlayQueueInNormalOrderObservable()
                .map(list -> mapList(list, this::toQueueItem));
    }

    private Observable<List<PlayQueueItem>> getPlayQueueInShuffledOrderObservable() {
        return playQueueDao.getPlayQueueInShuffledOrderObservable()
                .map(list -> mapList(list, this::toQueueItem));
    }

    private List<PlayQueueEntity> toEntityList(List<Composition> compositions,
                                               int position,
                                               int shuffledPosition) {
        List<PlayQueueEntity> entityList = new ArrayList<>(compositions.size());

        for (Composition composition: compositions) {
            PlayQueueEntity playQueueEntity = new PlayQueueEntity();
            playQueueEntity.setAudioId(composition.getId());
            playQueueEntity.setPosition(position++);
            playQueueEntity.setShuffledPosition(shuffledPosition++);

            entityList.add(playQueueEntity);
        }
        return entityList;
    }

    private PlayQueueItem toQueueItem(PlayQueueItemDto dto) {
        return new PlayQueueItem(dto.getItemId(), dto.getComposition());
    }
}

File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/play_queue/PlayQueueDaoWrapper.kt
package com.github.anrimian.musicplayer.data.database.dao.play_queue

import android.database.sqlite.SQLiteCantOpenDatabaseException
import androidx.sqlite.db.SimpleSQLiteQuery
import com.github.anrimian.musicplayer.data.database.LibraryDatabase
import com.github.anrimian.musicplayer.data.database.entities.play_queue.PlayQueueEntity
import com.github.anrimian.musicplayer.data.repositories.state.UiStateRepositoryImpl
import com.github.anrimian.musicplayer.domain.Constants
import com.github.anrimian.musicplayer.domain.models.composition.CompositionModel
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem
import com.github.anrimian.musicplayer.domain.utils.functions.Opt
import io.reactivex.rxjava3.core.Observable

/**
 * Created on 02.07.2018.
 */
class PlayQueueDaoWrapper(
    private val libraryDatabase: LibraryDatabase,
    private val playQueueDao: PlayQueueDao,
) {

    private var deletedItem: PlayQueueEntity? = null

    fun getPlayQueueObservable(
        isRandom: Boolean,
        useFileName: Boolean,
    ): Observable<List<PlayQueueItem>> {
        var query = PlayQueueDao.getCompositionQuery(useFileName)
        query += if (isRandom) "ORDER BY shuffledPosition" else "ORDER BY position"
        val sqlQuery = SimpleSQLiteQuery(query)
        return playQueueDao.getPlayQueueObservable(sqlQuery)
    }

    fun getWindowPlayQueueObservable(
        itemId: Long,
        isRandom: Boolean,
        useFileName: Boolean,
        startOffset: Int,
        endOffset: Int
    ): Observable<List<PlayQueueItem>> {
        val position = getPosition(itemId, isRandom) - startOffset//TODO-W handle negative value
        val itemsCount = endOffset + startOffset
        val query = PlayQueueDao.getWindowCompositionQuery(useFileName, isRandom, itemsCount)
        val sqlQuery = SimpleSQLiteQuery(query, arrayOf(position))
        return playQueueDao.getPlayQueueObservable(sqlQuery)
    }

    fun reshuffleQueue(currentItemId: Long) {
        libraryDatabase.runInTransaction {
            val ids = playQueueDao.getPlayQueueIds()
            if (ids.isEmpty()) {
                return@runInTransaction
            }
            ids.shuffle()

            val currentItemIndex = ids.indexOf(currentItemId)

            // If the current song is not already at the top (index 0),
            //  swap it with whatever song is at the top.
            if (currentItemIndex != -1 && currentItemIndex != 0) {
                val itemAtTop = ids[0]
                ids[0] = ids[currentItemIndex]
                ids[currentItemIndex] = itemAtTop
            }

            // prevent sqlite constraint error
            ids.forEachIndexed { newPosition, songId ->
                playQueueDao.updateShuffledPosition(songId, (newPosition + 1) * -1)
            }

            ids.forEachIndexed { newPosition, itemId ->
                playQueueDao.updateShuffledPosition(itemId, newPosition)
            }
        }
    }

    fun insertNewPlayQueue(
        compositionIds: List<Long>,
        randomPlayingEnabled: Boolean,
        startPosition: Int,
    ): Long {
        return libraryDatabase.runInTransaction<Long> {

            val shuffledPositions = compositionIds.indices.toMutableList()
            shuffledPositions.shuffle()

            val entities = compositionIds.mapIndexed { index, compositionId ->
                PlayQueueEntity(0, compositionId, index, shuffledPositions[index])
            }

            playQueueDao.deletePlayQueue()
            playQueueDao.insertItems(entities)

            return@runInTransaction if (randomPlayingEnabled) {
                val targetShuffledPosition = if (startPosition == Constants.NO_POSITION) {
                    0
                } else {
                    shuffledPositions[startPosition]
                }
                playQueueDao.getItemIdAtShuffledPosition(targetShuffledPosition)!!
            } else {
                val targetPosition = if (startPosition == Constants.NO_POSITION) {
                    0
                } else {
                    startPosition
                }
                playQueueDao.getItemIdAtPosition(targetPosition)!!
            }
        }
    }

    fun getItemObservable(id: Long, useFileName: Boolean): Observable<Opt<PlayQueueItem>> {
        var query = PlayQueueDao.getCompositionQuery(useFileName)
        query += "WHERE itemId = ? LIMIT 1"
        val sqlQuery = SimpleSQLiteQuery(query, arrayOf<Any>(id))
        return playQueueDao.getItemObservable(sqlQuery)
            .map { itemArray -> Opt(itemArray.firstOrNull()) }
    }

    fun deleteItem(itemId: Long) {
        deletedItem = playQueueDao.getItem(itemId)
        playQueueDao.deleteItem(itemId)
    }

    fun restoreDeletedItem(): Long? {
        if (deletedItem != null) {
            return playQueueDao.insertItem(deletedItem!!)
        }
        return null
    }

    fun swapItems(firstItem: PlayQueueItem, secondItem: PlayQueueItem, shuffleMode: Boolean) {
        libraryDatabase.runInTransaction {
            val firstId = firstItem.itemId
            val secondId = secondItem.itemId
            if (shuffleMode) {
                val firstPosition = playQueueDao.getShuffledPosition(firstId)
                val secondPosition = playQueueDao.getShuffledPosition(secondId)

                playQueueDao.updateShuffledPosition(secondId, Int.MIN_VALUE)
                playQueueDao.updateShuffledPosition(firstId, secondPosition)
                playQueueDao.updateShuffledPosition(secondId, firstPosition)
            } else {
                val firstPosition = playQueueDao.getPosition(firstId)
                val secondPosition = playQueueDao.getPosition(secondId)

                playQueueDao.updateItemPosition(secondId, Int.MIN_VALUE)
                playQueueDao.updateItemPosition(firstId, secondPosition)
                playQueueDao.updateItemPosition(secondId, firstPosition)
            }
        }
    }

    fun addCompositionsToEndQueue(compositions: List<CompositionModel>): Long {
        return libraryDatabase.runInTransaction<Long> {
            val positionToInsert = playQueueDao.getLastPosition() + 1
            val shuffledPositionToInsert = playQueueDao.getLastShuffledPosition() + 1
            val entities = toEntityList(compositions, positionToInsert, shuffledPositionToInsert)
            val ids = playQueueDao.insertItems(entities)
            return@runInTransaction ids[0]
        }
    }

    fun addCompositionsToQueue(compositions: List<CompositionModel>, currentItemId: Long): Long {
        return libraryDatabase.runInTransaction<Long> {
            var positionToInsert = 0
            var shuffledPositionToInsert = 0
            if (currentItemId != UiStateRepositoryImpl.NO_ITEM) {
                val currentPosition = playQueueDao.getPosition(currentItemId)
                val currentShuffledPosition = playQueueDao.getShuffledPosition(currentItemId)

                val increaseBy = compositions.size
                val lastPosition = playQueueDao.getLastPosition()
                for (pos in lastPosition downTo currentPosition + 1) {
                    playQueueDao.increasePosition(increaseBy, pos)
                }

                val lastShuffledPosition = playQueueDao.getLastShuffledPosition()
                for (pos in lastShuffledPosition downTo currentShuffledPosition + 1) {
                    playQueueDao.increaseShuffledPosition(increaseBy, pos)
                }

                positionToInsert = currentPosition + 1
                shuffledPositionToInsert = currentShuffledPosition + 1
            }

            val entities = toEntityList(compositions, positionToInsert, shuffledPositionToInsert)
            val ids = playQueueDao.insertItems(entities)
            return@runInTransaction ids[0]
        }
    }

    fun getPosition(id: Long, isShuffle: Boolean): Int {
        return if (isShuffle) {
            playQueueDao.getShuffledPosition(id)
        } else {
            playQueueDao.getPosition(id)
        }
    }

    fun getLastPosition(isShuffled: Boolean): Int {
        return if (isShuffled) {
            playQueueDao.getLastShuffledPosition()
        } else {
            playQueueDao.getLastPosition()
        }
    }

    fun getPositionObservable(id: Long, isShuffle: Boolean): Observable<Int> {
        val observable = if (isShuffle) {
            playQueueDao.getShuffledPositionObservable(id)
        } else {
            playQueueDao.getPositionObservable(id)
        }
        return observable
            .retry(
                DB_OBSERVABLE_RETRY_COUNT.toLong(),
                { t -> t is SQLiteCantOpenDatabaseException }
            )
            .distinctUntilChanged()
    }

    fun getIndexPositionObservable(id: Long, isShuffle: Boolean): Observable<Int> {
        val observable = if (isShuffle) {
            playQueueDao.getShuffledIndexPositionObservable(id)
        } else {
            playQueueDao.getIndexPositionObservable(id)
        }
        return observable.filter { pos -> pos >= 0 }
            .distinctUntilChanged()
    }

    fun getNextQueueItemId(currentItemId: Long, isShuffled: Boolean): Long {
        return if (isShuffled) {
            playQueueDao.getNextShuffledQueueItemId(currentItemId) ?: playQueueDao.getFirstShuffledItem()
        } else {
            playQueueDao.getNextQueueItemId(currentItemId) ?: playQueueDao.getFirstItem()
        }
    }

    fun getPreviousQueueItemId(currentItemId: Long, isShuffled: Boolean): Long {
        return if (isShuffled) {
            playQueueDao.getPreviousShuffledQueueItemId(currentItemId) ?: playQueueDao.getLastShuffledItem()
        } else {
            playQueueDao.getPreviousQueueItemId(currentItemId) ?: playQueueDao.getLastItem()
        }
    }

    fun getItemAtPosition(position: Int, isShuffled: Boolean): Long? {
        return if (isShuffled) {
            playQueueDao.getItemIdAtShuffledPosition(position)
        } else {
            playQueueDao.getItemIdAtPosition(position)
        }
    }

    fun deletePlayQueue() {
        playQueueDao.deletePlayQueue()
    }

    fun getPlayQueueSize() = playQueueDao.getPlayQueueSize()

    fun getPlayQueueSizeObservable() = playQueueDao.getPlayQueueSizeObservable()

    fun insertTrackPosition(itemId: Long, trackPosition: Long) {
        libraryDatabase.runInTransaction {
            val currentPosition = playQueueDao.getTrackPosition(itemId)
            if (currentPosition == trackPosition) {
                return@runInTransaction
            }
            playQueueDao.insertTrackPosition(itemId, trackPosition, System.currentTimeMillis())
            playQueueDao.deleteOldestTrackPosition()
        }
    }

    fun getTrackPosition(itemId: Long): Long {
        return playQueueDao.getTrackPosition(itemId)
    }

    private fun toEntityList(
        compositions: List<CompositionModel>,
        position: Int,
        shuffledPosition: Int,
    ): List<PlayQueueEntity> {
        return compositions.mapIndexed { index, composition ->
            PlayQueueEntity(0, composition.id, position + index, shuffledPosition + index)
        }
    }

    companion object {
        private const val DB_OBSERVABLE_RETRY_COUNT = 5
    }
}
